packages/@lwc/engine-dom/src/patches/global-registry.ts

```diff
@@ -0,0 +1,398 @@
+        );
+    }
+    (nativeRegistry as any)[registryPatchedSymbol] = true;
+
+    // patch for the global registry define mechanism
+    CustomElementRegistry.prototype.define = function define(
+        this: CustomElementRegistry,
+        tagName: string,
+        constructor: CustomElementConstructor,
+        options?: ElementDefinitionOptions
+    ): void {
+        if (options && options.extends) {
+            throw new DOMException(
+                'NotSupportedError: "extends" key in customElements.define() options is not supported.'
+            );
+        }
+        nativeGet.call(this, tagName); // SyntaxError if The provided name is not a valid custom element name.
+        tagName = tagName.toLowerCase();
+        if (!isUndefined(definitionsByTag.get(tagName))) {
+            throw new DOMException(
+                `Failed to execute 'define' on 'CustomElementRegistry': the name "${tagName}" has already been used with this registry`
+            );
+        }
+        const definition = getDefinitionForConstructor(constructor);
+        if (!isUndefined(definitionsByClass.get(constructor))) {
+            throw new DOMException(
+                `Failed to execute 'define' on 'CustomElementRegistry': this constructor has already been used with this registry`
+            );
+        }
+        definitionsByTag.set(tagName, definition);
+        definitionsByClass.set(constructor, definition);
+        let PivotCtor = pivotCtorByTag.get(tagName);
+        if (isUndefined(PivotCtor)) {
+            PivotCtor = createPivotingClass(tagName, definition);
+            // Register a pivoting class which will handle global registry initializations
+            nativeDefine.call(nativeRegistry, tagName, PivotCtor);
+        }
+        // For globally defined custom elements, the definition associated
+        // to the UserCtor has a back-pointer to PivotCtor in case the user
+        // new the UserCtor, so we know how to create the underlying element.
+        definition.PivotCtor = PivotCtor;
+        // Upgrade any elements created in this scope before define was called
+        // which should be exhibit by LWC using a tagName (in a template)
+        // before the same tagName is registered as a global, while others
+        // are already created and waiting in the global context, that will
+        // require immediate upgrade when the new global tagName is defined.
+        const awaiting = awaitingUpgrade.get(tagName);
+        if (!isUndefined(awaiting)) {
+            awaitingUpgrade.delete(tagName);
+            for (const element of awaiting) {
+                const registeredDefinition = pendingRegistryForElement.get(element);
+                if (!isUndefined(registeredDefinition)) {
+                    pendingRegistryForElement.delete(element);
+                    internalUpgrade(element, registeredDefinition, definition);
+                }
+            }
+        }
```

Comment:
@caridy I am having trouble thinking of a case where we will actually enter into this `if` block. I read the comment, but I don't quite understand. It looks like `awaitingUpgrade` is only ever populated if `connectedCallback`/`disconnectedCallback` is called for an element with no `definitionForElement`. But this is defined in `createPivotingClass`, and if we're inside of `createPivotingClass`, then we should have already created a `definitionForElement` via `internalUpgrade`.

I'm trying to think of a test to actually get the code to go down this code path, but I'm not able to, so far. Any ideas?

Reply:
I tried again to find a test for _all_ the unused code paths in this file. I wasn't able to cover everything, but out of an abundance of caution, I kept some of the original `if` checks and just added comments explaining why they were there.